src/lib/Parser.hs

```diff
@@ -260,6 +262,56 @@ topLet = do
         k = if v == mkName "eff" then EffectRowKind else TypeKind
         uTyKind = WithSrc (Just pos) $ UPrimExpr $ TCExpr k
 
+interfaceDef :: Parser [UDecl]
+interfaceDef = do
+  keyWord InterfaceKW
+  -- TODO: allow multi-parameter type classes
+  interfaceName <- upperName <|> symName
+  (oneAnnBinder, pos) <- withPos annBinder
+  keyWord WhereKW
+  record <- withSrc $ URecordTy <$> interfaceRecordFields ":"
+  let tyCon = UConDef interfaceName $ toNest [oneAnnBinder]
+  let consName = mkInterfaceConsName interfaceName
+  let varTypeName = case oneAnnBinder of
+        Bind v   -> varName v
+        -- This will fail during typechecking.
+        Ignore _ -> mkName "_"
+  let funDefs = mkFunDefs (pos, varTypeName, interfaceName) record
+  return $ (UData tyCon [UConDef consName (toNest [Ignore record])]) : funDefs
+  where
+    mkFunDefs :: (SrcPos, Name, Name) -> UExpr -> [UDecl]
+    mkFunDefs meta recordType
+      | (WithSrc _ (URecordTy (Ext (LabeledItems items) _))) <- recordType =
+          mkFunDefs' meta $ M.toList items
+      | otherwise = []
+      where
+        mkFunDefs' tup = (map (mkOneFunDef tup)) . concat . map expand
+        expand tup = map (\e -> (fst tup, e)) $ toList $ snd tup
```

Comment:
The `toList` here could have been `NE.head` right? I think that `mkFunDefs'` can just be this (and you can remove the `meta` argument, since it's still in scope):
```hs
mkFunDefs' = fmap (\(name, ty `NE.(:|)` []) -> mkOneFunDef meta (name, ty))
```